### ISSUE TITLE
[QUESTION] How to use to optimize NN hyperparameters

### ISSUE THREAD

**Issue by loretoparisi** [facebookresearch/nevergrad#1]
I'm currently using [hyperopt](https://github.com/hyperopt/hyperopt) to run an objective function to optimize hyperparameters for [FastText](https://github.com/facebookresearch/fastText) so the optimization we'll have hyperparameters (like learning rate, dimension, epoch, etc.) as input and going to optimize the resulting model with precision and recall. How to achieve the same using nevergrad?

```python
from fastText import train_supervised
from hyperopt import hp, fmin, tpe, space_eval, STATUS_OK

def objective_function(configuration, return_model=False):
    model = train_supervised(input=FILE_TRAIN, label=u'__label__', thread=12,
        epoch=configuration['epoch'],
        lr=configuration['learning_rate'],
        dim=configuration['dimension'],
        ws=configuration['window_size'],
        wordNgrams=configuration['word_ngrams']
        )
    samples_number, precision, recall = model.test(FILE_VALID, k=1)
    if return_model:
        return {'loss': -precision, 'status': STATUS_OK}, model
    return {'loss': -precision, 'status': STATUS_OK }

if __name__ == "__main__":
    space = {
        'epoch': hp.choice('epoch_', [10, 20, 30, 40, 50, 60, 70, 80]),
        'learning_rate': hp.choice('learning_rate_', [0.1, 0.3, 0.5, 0.7, 1]),
        'dimension': hp.choice('dimension_', [10, 20, 50, 70, 100, 150, 200, 250, 300]),
        'window_size': hp.choice('window_size_', range(5, 15, 1)),
        'word_ngrams': hp.choice('word_ngrams_', range(1, 5, 1))
    }

    # Optimization
    best = fmin(fn=objective_function, space=space, algo=tpe.suggest, max_evals=5) #max_evals=200

    # best model
    configuration_best = space_eval(space, best)
    print(datetime_now(), "Best (configuration): ", configuration_best)

    print(datetime_now(), "Training best config...")
    _, model = objective_function(configuration_best, return_model=True)
    model.save('model_opt.bin')

    print(datetime_now(), "Test set evaluation")
    samples_number, precision, recall = model.test(FILE_TEST, k=1)
    print_results(samples_number, precision, recall)
```

**Comment by jrapin** [comment]
The "Instrumentation" part of the README may be useful for this. The instrumentation tools we provide help you convert your function defined on any space into a function defined on R^n, which can then be fed into on optimizer. I would use `OrderedDiscrete` variables since you want to try a discrete set of ordered values.

```python
from nevergrad import instrumentation as instru
from nevergrad.optimization import optimizerlib

# argument transformation
epoch_var = instru.variables.OrderedDiscrete([10, 20, 30, 40, 50, 60, 70, 80])) 
learning_rate_var = ...
...

# the "instrumented" function, defined on R^n (with n=5 in this case)
func = instru.InstrumentedFunction(train_supervised,
                                   input=FILE_TRAIN,
                                   label=u'__label__',
                                   thread=12,
                                   epoch=epoch_var,
                                   lr=lr_var,
                                   dim=dim_var,
                                   ws=ws_var,
                                   wordNgrams=wordNgrams_var)

# a budget must be explicitely provided, there is no stopping criterion for now
budget = 100
# you can choose among many algorithm, check optimizerlib.registry
optimizer = optimizerlib.TwoPointsDE(dimension=func.dimension, budget=budget,
                                     num_workers=1)
# you can also provide an executor (see concurrent.futures)
# when num_workers > 1 for parallel computation
recommendation = optimizer.optimize(ifunc)

# The recommandation is provided in R^n and are tranforms of the parameters
# that interest you.
# You can tranform it back ang get the actual values you want with:
args, kwargs = ifunc.convert_to_arguments(recommendation)
```

The library is still under heavy development. The API will probably still evolve a lot and feedbacks are welcome if you have ideas on how to make it better/simpler ;)

**Comment by teytaud** [comment]
Due to popular request, we created a few examples about how to use Nevergrad for optimizing hyperparameters of models that can be conveniently trained from Python:
https://github.com/facebookresearch/nevergrad/blob/master/docs/machinelearning.md

Hope that helps, and comments very welcome!

**Comment by loretoparisi** [comment]
@teytaud great thank you!!!

**Comment by aqibsaeed** [comment]
To maximize precision/recall with NeverGrad, simply adding minus sign in front of the metric (-precision) will work or does it requires other changes as well?

**Comment by LucasCampos** [comment]
@teytaud The link for that part of the documentation is now broken. I believe it has been moved to https://github.com/facebookresearch/nevergrad/blob/master/docs/machinelearning.rst
